src/elm/external.ts

```diff
@@ -22,21 +25,48 @@ export class Retrieve extends Expression {
   }
 
   exec(ctx: Context) {
-    let records = ctx.findRecords(this.templateId != null ? this.templateId : this.datatype);
-    let codes = this.codes;
-    if (this.codes && typeof this.codes.exec === 'function') {
-      codes = this.codes.execute(ctx);
-      if (codes == null) {
+    // Object with retrieve information to pass back to patient source
+    // Always assign datatype. Assign codeProperty and dateProperty if present
+    const retrieveDetails: RetrieveDetails = {
+      datatype: this.datatype,
+      ...(this.codeProperty ? { codeProperty: this.codeProperty } : {}),
+      ...(this.dateProperty ? { dateProperty: this.dateProperty } : {})
+    };
+
+    if (this.codes) {
+      const resolvedCodes: Code[] | ValueSet | undefined = this.codes.execute(ctx);
+
+      if (resolvedCodes == null) {
         return [];
       }
+
+      retrieveDetails.codes = resolvedCodes;
     }
-    if (codes) {
-      records = records.filter((r: any) => this.recordMatchesCodesOrVS(r, codes));
-    }
+
     // TODO: Added @dateProperty check due to previous fix in cql4browsers in cql_qdm_patient_api hash: ddbc57
-    if (this.dateRange && this.dateProperty) {
-      const range = this.dateRange.execute(ctx);
-      records = records.filter((r: any) => range.includes(r.getDateOrInterval(this.dateProperty)));
+    if (this.dateRange) {
+      const range: Interval = this.dateRange.execute(ctx);
+
+      retrieveDetails.dateRange = range;
```

Comment:
I didn't check to make sure tsc is happy, but I assume this can be simplified to a single statement...
```suggestion
      retrieveDetails.dateRange = this.dateRange.execute(ctx);
```